src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,377 @@
+            }
+
+            return targetType;
+        }
+
+        private static bool TrySetTargetMethod(ref GenerationContext context, string name)
+        {
+            TypeDesc targetType = context.TargetType;
+
+            TypeDesc declTypeDesc;
+            TypeDesc maybeTypeDesc;
+            foreach (MethodDesc md in targetType.GetMethods())
+            {
+                // Check the target and current method match static/instance state.
+                if (context.IsTargetStatic != md.Signature.IsStatic)
+                {
+                    continue;
+                }
+
+                // Check for matching name
+                if (!md.Name.Equals(name))
+                {
+                    continue;
+                }
+
+                // Validate calling convention.
+                if ((MethodSignatureFlags.UnmanagedCallingConventionMask & md.Signature.Flags)
+                    != (MethodSignatureFlags.UnmanagedCallingConventionMask & context.Declaration.Signature.Flags))
+                {
+                    continue;
+                }
+
+                // Validate the return type and prepare for validating
+                // the current signature's argument list.
+                int sigCount = context.Declaration.Signature.Length;
+                if (context.Kind == UnsafeAccessorKind.Constructor)
+                {
+                    if (!md.Signature.ReturnType.IsVoid)
+                    {
+                        continue;
+                    }
+                }
+                else
+                {
+                    declTypeDesc = context.Declaration.Signature.ReturnType;
+                    maybeTypeDesc = md.Signature.ReturnType;
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        continue;
+                    }
+
+                    // Non-constructor accessors skip the first argument
+                    // when validating the target argument list.
+                    sigCount--;
+                }
+
+                // Validate argument count matches.
+                if (sigCount != md.Signature.Length)
+                {
+                    continue;
+                }
+
+                // Validate arguments match - reverse order
+                for (; sigCount > 0; --sigCount)
+                {
+                    declTypeDesc = context.Declaration.Signature[sigCount];
+                    maybeTypeDesc = md.Signature[sigCount];
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        break;
+                    }
+                }
+
+                // If we validated all arguments, we have a match.
+                if (sigCount != 0)
+                {
+                    continue;
+                }
+
+                context.TargetMethod = md;
+                return true;
+            }
+            return false;
```

Comment:
Should we try looking in the base type for non-constructor methods?

Moving methods to a base type is a non-breaking change in IL terms.

Reply:
I don't think this feature needs that level of support. I could be wrong about use case, but unless there is a specific ask I am inclined to require callers to be explicit about their needs and we avoid attempting to do more complicated lookup.

Reply:
IIRC we did changes like this in the past so this could happen if a source generator source generates against netstandard 2.0 reference assemblies and we see .NET 8 at runtime. I don't have any examples but it is an allowed change:

https://github.com/dotnet/runtime/blob/20ca966b59c96ddbc8ce6b2c6f415d9c93f2c49a/docs/coding-guidelines/breaking-change-rules.md?plain=1#L180

It will also happen if unsafe accessors are used in a NuGet and target a different NuGet that later does the (normally) non-breaking change that breaks them if we don't do it.

We can wait for feedback, but wanted to bring it up.